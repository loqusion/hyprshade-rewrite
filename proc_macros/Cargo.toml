[package]
name = "proc_macros"
version = "0.1.0"
publish = false
edition = "2021"

[dependencies]
lazy_static = "1.5.0"
quote = "1.0.36"
syn = "2.0.74"

[lib]
proc-macro = true
