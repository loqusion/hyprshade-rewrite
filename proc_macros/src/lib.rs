use std::env;

use lazy_static::lazy_static;
use proc_macro::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input, ItemFn,
};

lazy_static! {
    static ref IS_HYPRLAND_RUNNING: bool = env::var_os("HYPRLAND_INSTANCE_SIGNATURE").is_some();
}

/// Attribute macro declaring a test function that will be ignored if the test is not running in a
/// Hyprland instance.
///
/// In other words, a test not running in a Hyprland instance should be roughly equivalent to the
/// following:
///
/// ```no_run
/// #[test]
/// #[ignore = "requires hyprland"]
/// fn test() { /* ... */ }
/// ```
///
/// **WARNING:** Once tests have been run, the decision to ignore tests will not be re-evaluated
/// until recompilation is triggered, e.g. by changing a file. This can lead to tests starting to
/// fail if the environment changes between runs.
#[proc_macro_attribute]
pub fn hyprland_test(attr: TokenStream, item: TokenStream) -> TokenStream {
    let item = parse_macro_input!(item as ItemFn);
    // Verify there are no attribute arguments
    let _ = parse_macro_input!(attr as HyprlandTestAttribute);

    let ignore_attr = if *IS_HYPRLAND_RUNNING {
        None
    } else {
        Some(quote! {
            #[ignore = "requires hyprland"]
        })
    };

    let expanded = quote! {
        #[test]
        #ignore_attr
        #item
    };

    TokenStream::from(expanded)
}

#[derive(Debug, Clone)]
struct HyprlandTestAttribute {}

impl Parse for HyprlandTestAttribute {
    fn parse(_input: ParseStream) -> syn::Result<Self> {
        Ok(Self {})
    }
}
