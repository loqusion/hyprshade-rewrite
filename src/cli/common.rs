use std::{collections::HashMap, convert::Infallible, str::FromStr};

pub const SHADER_HELP: &str = "Which shader to turn on";
pub const SHADER_HELP_LONG: &str = "\
    Which shader to turn on\n\
\n\
    May be a name (e.g. `blue-light-filter`)\n\
    or a path (e.g. `~/.config/hypr/shaders/blue-light-filter.glsl`)\
";

#[derive(Debug, Clone)]
pub struct ArgVar(String);

impl FromStr for ArgVar {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self(s.to_string()))
    }
}

impl TryFrom<ArgVar> for HashMap<String, mustache::Data> {
    type Error = ArgVarError;

    fn try_from(value: ArgVar) -> Result<Self, Self::Error> {
        let ArgVar(s) = value;
        match s.split('=').collect::<Vec<_>>()[..] {
            ["", _] => Err(ArgVarError::EmptyKey(s)),
            [_, ""] => Err(ArgVarError::EmptyValue(s)),
            [dotted_key, value] => {
                let mut key_split = dotted_key.split('.').rev();
                let mut map: HashMap<String, mustache::Data> = HashMap::from([(
                    key_split.next().unwrap().to_string(),
                    mustache::Data::String(value.to_string()),
                )]);

                for key in key_split {
                    map = HashMap::from([(key.to_string(), mustache::Data::Map(map))]);
                }
                Ok(map)
            }
            [_, _, ..] => Err(ArgVarError::TooManyEquals(s)),
            [_] => Err(ArgVarError::NoEquals(s)),
            [] => Err(ArgVarError::NoContent),
        }
    }
}

impl FromIterator<ArgVar> for Result<mustache::Data, Vec<ArgVarError>> {
    fn from_iter<T: IntoIterator<Item = ArgVar>>(iter: T) -> Self {
        let mut map: HashMap<String, mustache::Data> = HashMap::new();

        let (things, errors): (Vec<_>, Vec<_>) = iter
            .into_iter()
            .map(|arg_var| arg_var.try_into())
            .partition(Result::is_ok);
        let things: Vec<_> = things.into_iter().map(Result::unwrap).collect();
        let mut errors: Vec<_> = errors.into_iter().map(Result::unwrap_err).collect();

        let merge_errors = things
            .into_iter()
            .filter_map(|data| ArgVar::deep_merge(&mut map, data).err())
            .collect::<Vec<_>>();
        errors.extend(merge_errors);

        if !errors.is_empty() {
            return Err(errors);
        }

        Ok(mustache::Data::Map(map))
    }
}

impl ArgVar {
    fn deep_merge(
        map: &mut HashMap<String, mustache::Data>,
        data: HashMap<String, mustache::Data>,
    ) -> Result<(), ArgVarError> {
        ArgVar::deep_merge_impl(map, data, &Vec::new())
    }

    fn deep_merge_impl(
        map: &mut HashMap<String, mustache::Data>,
        data: HashMap<String, mustache::Data>,
        ctx: &[&str],
    ) -> Result<(), ArgVarError> {
        use std::collections::hash_map::Entry;

        for (key, value) in data {
            match map.entry(key) {
                Entry::Vacant(entry) => {
                    entry.insert(value);
                }
                Entry::Occupied(mut entry) => {
                    let key = entry.key().clone();
                    let entry_value = entry.get_mut();
                    match (entry_value, value) {
                        (mustache::Data::Map(inner_map), mustache::Data::Map(inner_value)) => {
                            let mut ctx = Vec::from(ctx);
                            ctx.push(&key);
                            ArgVar::deep_merge_impl(inner_map, inner_value, &ctx)?;
                        }
                        (mustache::Data::Map(inner_map), value) => {
                            dbg!(&ctx);
                            dbg!(&inner_map);
                            dbg!(&value);
                            return Err(ArgVarError::Merge);
                        }
                        (entry_value, mustache::Data::Map(inner_value)) => {
                            dbg!(&ctx);
                            dbg!(&entry_value);
                            dbg!(&inner_value);
                            return Err(ArgVarError::Merge);
                        }
                        (entry_value, value) => {
                            dbg!(&ctx);
                            dbg!(&entry_value);
                            dbg!(&value);
                            return Err(ArgVarError::Merge);
                        }
                    }
                }
            }
        }

        Ok(())
    }
}

#[derive(Debug, thiserror::Error)]
pub enum ArgVarError {
    #[error("{0:?}: empty key (variable must be in the form `key=value`)")]
    EmptyKey(String),
    #[error("{0:?}: empty value (variable must be in the form `key=value`)")]
    EmptyValue(String),
    #[error("{0:?}: too many `=` (variable must be in the form `key=value`)")]
    TooManyEquals(String),
    #[error("{0:?}: missing `=` (variable must be in the form `key=value`)")]
    NoEquals(String),
    #[error("missing (variable must be in the form `key=value`)")]
    NoContent,
    #[error("encountered errors when merging")]
    Merge,
}

#[allow(clippy::manual_try_fold)]
pub fn arg_vars_to_data<I>(iter: I, opt_name: &str) -> eyre::Result<mustache::Data>
where
    I: IntoIterator<Item = ArgVar>,
{
    use color_eyre::Section;
    use eyre::eyre;

    iter.into_iter().collect::<Result<_, _>>().or_else(|errs| {
        errs.into_iter().fold(
            Err(eyre!("encountered errors parsing --{opt_name}")),
            |report, e| report.error(e),
        )
    })
}
