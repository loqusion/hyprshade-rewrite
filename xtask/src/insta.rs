use std::process::ExitCode;

use xshell::{cmd, Shell};

use crate::{hooks::RestoreShaderHook, test::CARGO_TEST_FLAGS};

pub fn main(shell: Shell, args: &[String]) -> eyre::Result<ExitCode> {
    let hook = RestoreShaderHook::new();

    let result = match args {
        [subcommand, args @ ..] if subcommand == "test" => {
            cmd!(shell, "cargo insta test {CARGO_TEST_FLAGS...} {args...}").run()
        }
        args => cmd!(shell, "cargo insta {args...}").run(),
    };

    hook.after();
    result?;

    Ok(ExitCode::SUCCESS)
}
