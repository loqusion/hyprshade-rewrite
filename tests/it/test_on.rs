use insta::assert_snapshot;
use proc_macros::hyprland_test;

use crate::common::{hyprshade_cmd_snapshot, CommandExt, Space};

#[hyprland_test]
fn builtin_shader() {
    let space = Space::new();
    let _stash = space.stash_runtime_shader("vibrance");
    space.hyprshade_cmd().arg("off").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("on").args(["vibrance"]), @r###"
    success: true
    exit_code: 0
    ----- stdout -----

    ----- stderr -----
    "###);

    assert_snapshot!(space.read_runtime_shader("vibrance"));
}

#[hyprland_test]
fn path_shader() {
    let space = Space::new();
    space.hyprshade_cmd().arg("off").run();
    let fixture_path = space.fixture_simple();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("on").arg(&fixture_path), @r###"
    success: true
    exit_code: 0
    ----- stdout -----

    ----- stderr -----
    "###);
}

mod error {
    use crate::common::{hyprshade_cmd_snapshot, Space};

    #[test]
    fn fails_resolving_shader() {
        let space = Space::new();
        hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("on").args(["invalid"]), @r###"
        success: false
        exit_code: 1
        ----- stdout -----

        ----- stderr -----
        Error: 
           0: [91mshader named "invalid" not found[0m

        Location:
           [LOCATION]
        "###);
    }

    #[test]
    fn fails_merging_var() {
        let space = Space::new();
        hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("on").args([
            "vibrance",
            "--var",
            "strength=0.5",
            "--var",
            "strength=0.6",
        ]), @r###"
        success: false
        exit_code: 2
        ----- stdout -----

        ----- stderr -----
        error: the argument '--var strength=0.6' cannot be used with '--var strength=0.5'

          tip: '--var strength=0.6' would override '--var strength=0.5'

        For more information, try '--help'.
        "###);
    }

    #[test]
    fn fails_parsing_var() {
        let space = Space::new();
        const INVALID: &[&[u8]] = &[
            b"",
            b"strength",
            b"=0.3",
            b"strength=",
            b"=strength=",
            b"strength=0.3=0.4",
            b"balance..red=0.3",
            b".balance.red=0.3",
            b"balance.red.=0.3",
            b"\xFF\xFF\xFF",
        ];
        for invalid in INVALID.iter() {
            hyprshade_cmd_snapshot!(
                String::from_utf8_lossy(invalid).as_ref(),
                space
                    .hyprshade_cmd()
                    .arg("on")
                    .args(["vibrance", "--var"])
                    .arg(unsafe { std::ffi::OsStr::from_encoded_bytes_unchecked(invalid) })
            );
        }
    }
}
