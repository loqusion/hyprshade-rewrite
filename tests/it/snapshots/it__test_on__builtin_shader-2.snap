---
source: tests/it/test_on.rs
expression: "space.read_runtime_shader(\"vibrance\")"
---
/*
 * Vibrance
 *
 * Enhance color saturation.
 * Also supports per-channel multipliers.
 *
 * Source: https://github.com/hyprwm/Hyprland/issues/1140#issuecomment-1614863627
 */

#version 300 es
precision highp float;

in vec2 v_texcoord;
uniform sampler2D tex;
out vec4 fragColor;

// see https://github.com/CeeJayDK/SweetFX/blob/a792aee788c6203385a858ebdea82a77f81c67f0/Shaders/Vibrance.fx#L20-L30

const vec3 Balance = vec3(
    float(1),
    float(1),
    float(1)
);

const float Strength = float(0.15);

const vec3 VIB_coeffVibrance = Balance * -Strength;

void main() {
    vec4 pixColor = texture(tex, v_texcoord);
    vec3 color = pixColor.rgb;

    vec3 VIB_coefLuma = vec3(0.212656, 0.715158, 0.072186);
    float luma = dot(VIB_coefLuma, color);

    float max_color = max(color.r, max(color.g, color.b));
    float min_color = min(color.r, min(color.g, color.b));
    float color_saturation = max_color - min_color;

    vec3 p_col = (sign(VIB_coeffVibrance) * color_saturation - 1.0) * VIB_coeffVibrance + 1.0;

    vec3 adjustedColor = mix(vec3(luma), color, p_col);
    fragColor = vec4(adjustedColor, pixColor.a);
}

// vim: ft=glsl
